### PR TITLE
KTOR-3325 Queue refresh token requests

### PR DESCRIPTION
**Pull request by rsinukov**
https://youtrack.jetbrains.com/issue/KTOR-3325

### CODE REVIEW

**Comment by e5l** [comment]
Please check native build

**Comment by e5l** [comment]
@andreyaksenov could you check API docs?

**Review by e5l — APPROVED** [review]

File: ktor-client/ktor-client-plugins/ktor-client-auth/common/src/io/ktor/client/plugins/auth/providers/BearerAuthProvider.kt [hunk 1]
```diff
@@ -10,7 +10,8 @@ import io.ktor.client.request.*
 import io.ktor.client.statement.*
 import io.ktor.http.*
 import io.ktor.http.auth.*
-import kotlin.jvm.*
+import kotlinx.atomicfu.*
+import kotlinx.coroutines.*
 
 /**
  * Add [BearerAuthProvider] to client [Auth] providers.
```
Comment (andreyaksenov) [inline ktor-client/ktor-client-plugins/ktor-client-auth/common/src/io/ktor/client/plugins/auth/providers/BearerAuthProvider.kt:17]: > Adds [BearerAuthProvider] to the client's [Auth] providers.

File: ktor-client/ktor-client-plugins/ktor-client-auth/common/src/io/ktor/client/plugins/auth/providers/BearerAuthProvider.kt [hunk 2]
```diff
@@ -45,10 +46,17 @@ public class BearerAuthConfig {
 
     public var realm: String? = null
 
+    /**
+     * Configures callback that refreshes token when 401 status code is received.
```
Comment (andreyaksenov) [inline ktor-client/ktor-client-plugins/ktor-client-auth/common/src/io/ktor/client/plugins/auth/providers/BearerAuthProvider.kt:50]: > Configures a callback that refreshes a token when the 401 status code is received.

File: ktor-client/ktor-client-plugins/ktor-client-auth/common/src/io/ktor/client/plugins/auth/providers/BearerAuthProvider.kt [hunk 3]
```diff
@@ -45,10 +46,17 @@ public class BearerAuthConfig {
...
+     */
     public fun refreshTokens(block: suspend RefreshTokensParams.() -> BearerTokens?) {
         _refreshTokens = block
     }
 
+    /**
+     * Configures callback that loads a cached token from a local storage.
```
Comment (andreyaksenov) [inline ktor-client/ktor-client-plugins/ktor-client-auth/common/src/io/ktor/client/plugins/auth/providers/BearerAuthProvider.kt:57]: > Configures a callback

**Review by andreyaksenov — APPROVED** [review]

**Pull request merged**